src/Juvix/Compiler/Internal/Translation/FromInternal/Analysis/Positivity/Checker.hs

```diff
@@ -28,157 +22,202 @@ type CheckPositivityEffects r =
      ]
     r
 
+data CheckPositivityArgs = CheckPositivityArgs
+  { _checkPositivityArgsInductive :: InductiveDef,
+    _checkPositivityArgsConstructorName :: Name,
+    _checkPositivityArgsInductiveName :: Name,
+    _checkPositivityArgsRecursionLimit :: Int,
+    _checkPositivityArgsErrorReference :: Maybe Expression,
+    _checkPositivityArgsTypeOfConstructor :: Expression
+  }
+
+makeLenses ''CheckPositivityArgs
+
 checkPositivity ::
   forall r.
   (CheckPositivityEffects r) =>
   InductiveDef ->
   Sem r ()
 checkPositivity ty = do
-  let indName = ty ^. inductiveName
-  numInductives <- HashMap.size <$> asks (^. infoInductives)
-  noCheckPositivity <- asks (^. E.entryPointNoPositivity)
-  forM_ (ty ^. inductiveConstructors) $ \ctor -> do
-    let ctorName = ctor ^. inductiveConstructorName
-        args = constructorArgs (ctor ^. inductiveConstructorType)
-    unless (noCheckPositivity || ty ^. inductivePositive) $
-      forM_
-        args
-        (checkStrictlyPositiveOccurrences ty ctorName indName numInductives Nothing)
+  unlessM (asks (^. E.entryPointNoPositivity)) $
+    forM_ (ty ^. inductiveConstructors) $ \ctor -> do
+      unless (ty ^. inductivePositive) $ do
+        numInductives <- HashMap.size <$> asks (^. infoInductives)
+        mapM_
+          ( \typeOfConstr ->
+              checkStrictlyPositiveOccurrences
+                ( CheckPositivityArgs
+                    { _checkPositivityArgsInductive = ty,
+                      _checkPositivityArgsConstructorName =
+                        ctor ^. inductiveConstructorName,
+                      _checkPositivityArgsInductiveName = ty ^. inductiveName,
+                      _checkPositivityArgsRecursionLimit = numInductives,
+                      _checkPositivityArgsErrorReference = Nothing,
+                      _checkPositivityArgsTypeOfConstructor = typeOfConstr
+                    }
+                )
+          )
+          (constructorArgs (ctor ^. inductiveConstructorType))
```

Comment:
When the mapped function takes multiple lines I usually prefer to use `forM_` inetead.
```suggestion
        forM_
          (constructorArgs (ctor ^. inductiveConstructorType))
          $ \typeOfConstr ->
              checkStrictlyPositiveOccurrences
                ( CheckPositivityArgs
                    { _checkPositivityArgsInductive = ty,
                      _checkPositivityArgsConstructorName =
                        ctor ^. inductiveConstructorName,
                      _checkPositivityArgsInductiveName = ty ^. inductiveName,
                      _checkPositivityArgsRecursionLimit = numInductives,
                      _checkPositivityArgsErrorReference = Nothing,
                      _checkPositivityArgsTypeOfConstructor = typeOfConstr
                    }
                )
```